### ISSUE TITLE
How to appropriately refactor code in A-MoreOOP

### ISSUE THREAD

**Issue by jyztintan** [nus-cs2103-AY2324S2/forum#126]
When refactoring our code to achieve A-MoreOOP, we had to create many new classes and integrate them into the main logic of our program. I realised that as I started refactoring it, creating new classes, it kind of depended on other new classes as well and my entire code broke until i successfully finished refactoring the code in all (old and new) classes. 

As a result, this was just one commit with many many changes. I feel like this is not the "correct" way to do it but i am not sure. Should we be instead creating classes one by one and integrating it into our main code? However, I think this would also have a lot of unnecessary refactoring as you go through each iteration.

**TLDR:**
1. What are the best practices for managing dependencies and smoothly integrating new classes with existing ones during a refactor, to keep the codebase functional throughout the process

2. In this (depth-first?) refactoring approach where new dependencies are introduced, is it advisable to commit changes even if the code is temporarily non-functional, or should commits only be made once the code is stable?

**Comment by tankh99** [comment]
1. I think that for best practices, avoid integrating the entire thing at once and integrate module by module first. Doing so would allow you to isolate any possible issues and debug that particular module first. 
2. I'm not sure about this, but I think code should preferably by functional when committing. But perhaps it's possible to avoid this issue entirely if you break your commits cleanly into smaller commits?

**Comment by itstrueitstrueitsrealitsreal** [comment]
I think its fine to make granular commits, you can just use stubs or comments with the todo keyword to highlight to the reviewer or yourself in the future that the methods are incomplete, just like the section in [this week's topics](https://nus-cs2103-ay2324s2.github.io/website/schedule/week3/topics.html).

Hope I helped!

**Comment by NeoHW** [comment]
hi! something that helped me was to first plan out the main overview of functionalities and which they all belonged to, then refactor based on just that package/grouping first !

**Comment by damithc** [comment]
This is a good point for discussion @tankh99 
Let's wait to see what others think, especially those who have done `A-MoreOOP` already.

**Comment by UdhayaShan1** [comment]
Other than the **is-a** relationships which can be clearly put as inheritances, the rest can be broken down to **has-a**, main class has a listener class that has a command parser which then has a tasklist modifier class... 
ragmented but alot clearer for me to see whats going on

**Comment by wilsonwid** [comment]
> 1. I think that for best practices, avoid integrating the entire thing at once and integrate module by module first. Doing so would allow you to isolate any possible issues and debug that particular module first. 

I think I agree with @tankh99's comment. In my opinion, this would also be more in line with the breadth-first iterative style we follow in this module.

**Comment by ys112** [comment]
Thanks for bringing forth the discussion, sharing my two cents:

1. It is good to firstly figure out the relationship between the classes (if you prefer visualisation, UML diagram would greatly help) and what are the classes supposed to do. Then, I would make a note on the chunk of existing codes according to where it would be refactored to and then shift it based on the modularity of the program, for e.g. if codes based on storage class is first refactored, then introduce storage only into the refactored duke bot. So, it would be always working line by line where new modules refactored are gradually put into the refactored duke bot.

2. I would prefer to push commit that is in a functional state whether up to a certain point or fully functional but not one that would fail upon running. Reason being that if I were to reverse back to an old commit due to certain issues, me and other developers working together on the same project would have to identify which are the working codes and commits.

**Comment by qinxutan** [comment]
hi! thank you for opening up this discussion. regarding your questions,

1. my train of thoughts while doing the refactoring was to group the various methods in the classes (e.g. all the methods to deal with user inputs in a group, all the methods to save user inputs into the txt in a grp etc). there after i will figure out the relationship between these classes and their inheritance (eg todo inherits from tasks etc). i would say a large part of it was making the main class a class with the "summary" of "category" classes and methods! so i was working from small classes and then removing the methods from the main class and then working my way up.

2. i only made the commit till the code is completely functional! like the other classmate mentioned, it is so that it would be easy for me to revert back to my last commit should my next change not run.

hope this helps!

**Comment by sun-ruiheng** [comment]
That's a great question that I also struggled with. What I did for planning was take a blank sheet of paper and list down all the classes I intended to introduce. Then, I looked through my very long Duke.java file full of messy functions and long if-else loops, and allocated each part of the code to its appropriate class on my plan. **Once you begin this process, it flows quite naturally** because each added allocation quite intuitively drives where the next part of the code should be. The is-a / has-a relationships also surface through this process.

When coding them out, I actually prefer to write code for the "highest level" classes first (e.g. Duke, TaskList), sort of defining the API, before getting the detailed smaller classes (like Storage, Task) to fulfil the API. This ensures you have full control over the structure of your program from the start. I do know people that prefer the other way, though.

**Comment by sun-ruiheng** [comment]
That's exactly an issue I grappled with, because the way I prefer to refactor is not very easy to track, as you mentioned.

When coding them out, I actually prefer to write code for the "highest level" classes first (e.g. Duke, TaskList), sort of defining the API, before getting the detailed smaller classes (like Storage, Task) to fulfil the API. This ensures you have full control over the structure of your program from the start. However, this runs into the exact problem you are describing.

The safer approach might be to work ground-up by creating the smaller classes first. You would need to have a clear plan of which methods / functionalities go into which classes, very clearly defining the I/O offered by each class. Then, begin with the smaller classes that don't depend on others, so you're adding new classes such as Storage while keeping your main long Duke file the same as before — throughout the process, your newly created classes are used minimally, if at all. Then once the smaller classes are done, you can refactor your main Duke class to incorporate their use. This way, there exists working code every step.

**Comment by hjungwoo01** [comment]
1. When refactoring and adding new classes, I integrated them one at a time. It helped keep things working smoothly without causing too much disruption
2. In the depth-first refactoring approach with new dependencies, I think its usually better to commit changes once the code is stable rather than when it's temporarily not working

**Comment by appleraincoat** [comment]
hi, just to share my experience, i did the same thing as you, whereby i only made 1 very big commit with all of the changes. 
to avoid confusion, i made a detailed high level diagram of all of the classes in my refactored version before actually coding, so i wouldn't get lost during the coding itself. it helped to a certain extent, but i feel that when possible, doing it in pieces (eg 1 class at a time) is definitely more ideal especially when the project classes do not rely on each other too heavily yet.

**Comment by jyztintan** [comment]
Hi everyone, 

Thank you for all your valuable insights. 

I just did a quick summary of the key takeaways from this discussion:

**Modular Integration**: Adopt a module-by-module approach to integration, as recommended by @tankh99 and @wilsonwid. This helps to isolate and debug issues more efficiently.

**Granular Commits:** Make smaller, more frequent commits. This ensures the codebase remains functional and facilitates easier debugging, as mentioned by @itstrueitstrueitsrealitsreal and @ys112.

**Planning and Visualization**: Before coding, plan out functionalities and their relationships. This approach, highlighted by @NeoHW and @qinxutan, provides a clear roadmap for refactoring. We can also draw high level diagrams and map out the various has-a, is-a relationships between the classes as suggested by @appleraincoat and @UdhayaShan1 

Maintain Functional Code: Aim to keep the codebase functional at every step of the refactoring process, as emphasized by @sun-ruiheng and @hjungwoo01.


TLDR: Best practices seem to include modular integration, making step-by-step commits, strategic planning, while maintaining functional code. 

Thanks everyone for the collective wisdom! I'll be marking this issue as resolved.

**Comment by damithc** [comment]
Good discussion everyone! To add a few points of my own,
1. Figuring out how to go from the current design to a new design in small self-contained non-breaking steps is bit of an art, and may not be obvious at first. It is not surprising that you'll find it hard when doing it for the first time. With experience, you'll get better at it. In fact, after having done it in whichever way, you can look back and think how would you have done it if you were to do it again. That kind of reflection will help you get better at it faster. I guess you are doing exactly that, by raising this issue. So, Kudos!
2. It is not necessary to know the target design when you start out to improve the current design. You can just make one small design improvement at a time (e.g., hmm... this method doesn't seem to fit in this class; let's extract it out to a new class). That way, every step will give you a slightly better design, without the burden of having to plan a major design change end-to-end. Related: the 'always keep the marshmallow on top' analogy I mentioned in week 3 briefing. 
3.  Rewriting is not the same as refactoring. The topic of refactoring is scheduled to be covered in a future week, at which point you can recall what you did at A-MoreOOP and reflect how much of it was refactoring and how much was rewriting.
